enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,218 @@
+
+Another benefit is that this approach should not break downgrades once CSI Migration becomes GA. In OCP, a downgrade is fundamentally a regular upgrade to an older version.
+That means that CVO downgrades components in the same order as it upgrades them: control-plane first, nodes later. This would impose an issue if the user downgraded from a version with CSI Migration enabled to a version with CSI Migration disabled. With the above patch in the downgraded version, that would not be a problem.
+
+It is important to notice that, with this carry-patch in OCP, Attach Detach Controller will _not_ change its current behaviour as long as nodes are not migrated to CSI, which is the default behaviour in OCP 4.8.
+
+#### Graduation
+
+During Tech Preview in OCP 4.8, users can enable the CSI migration using a *FeatureSet*. It could be shared with [CCMO](https://github.com/openshift/enhancements/pull/463), but we may want to have a specific *FeatureSet* only for CSI Migration for users that want to enable CSI Migration without migrating to an external cloud provider.
+
+Once CSI migration reaches GA in upstream, the associated feature gates will be enabled by default. As a result, it will remove the carry-patch above. Also, users will not have to use *FeatureSets* anymore.
+
+### Risks and Mitigations
+
+A carry-patch means that OCP will be the only Kubernetes distribution exercising this code path, which can lead us to bugs that were not seen anywhere. However, we are confident that the patch is small and self-contained enough to be used.
+
+## Design Details
+
+### Test Plan
+
+#### E2E jobs
+
+We want E2E jobs for all migrated plugins ready at **Tech Preview** time.
+
+For each E2E job:
+
+1. Install an OCP cluster.
+1. Enable the `CSIMigration` _FeatureSet_.
+1. Run E2E tests for in-tree volume plugins. This should use the CSI driver instead.
+1. Disable the `FeatureSet`.
+1. Once again, run E2E tests for in-tree volume plugins.
+
+In addition to that, as a stretch goal, we want a separate job that:
+
+1. Runs a `StatefulSet`.
+1. Enables the migration _FeatureSets_.
+1. Wait for all components to have the right feature flags.
+1. Checks if the StatefulSet survives.
+
+Once CSI migration is GA, we expect the regular upgrade jobs will cover upgrades from an OCP version with migration disabled to a version with migration enabled.
+
+### Graduation Criteria
+
+Each storage plugin will be migrated at their own paces and will have different support phases across OCP releases. For instance, GCP and Cinder migration would be Tech Preview and OCP 4.8 and GA in 4.9. AWS EBS will _most likely_ be Tech Preview in 4.9 and GA in 4.10.
+
+Having said that, the items below only cover the initial targets for CSI Migration: GCP PD and Cinder.
+
+1. Tech Preview in 4.8:
+
+* Introduce a new *FeatureSet* in openshift/api called `CSIMigration`. This step only needs to be done once because the same object will contain all
```

Comment:
This is an incomplete thought:

> ...the same object will contain all

Contain all CSIMigration flags, such as _CSIMigrationGCE_ and _CSIMigrationCinder_ ?

Reply:
Dropped sentence.